nengo_fpga/tests/test_id.py

```diff
@@ -0,0 +1,348 @@
+    extractor.connect_ssh_client(ssh_user, remote_ip)
+
+    # Create the expected call based on the config
+    if ssh_method is None:
+        connect_mock.assert_called_once_with(
+            remote_ip, port=ssh_port, username=ssh_user
+        )
+    elif ssh_method[0] == "ssh_pwd":
+        connect_mock.assert_called_once_with(
+            remote_ip, port=ssh_port, username=ssh_user, password=ssh_method[1]
+        )
+    elif ssh_method[0] == "ssh_key":
+        connect_mock.assert_called_once_with(
+            remote_ip, port=ssh_port, username=ssh_user, key_filename=ssh_method[1]
+        )
+
+
+def test_connect_thread_func(dummy_extractor, dummy_com, config_contents, mocker):
+    """Test the IDExtractor's connect_thread_func function
+
+       Similar to the test in "test_id"
+    """
+
+    # Don't use ssh connections
+    ssh_client_mock = mocker.patch.object(dummy_extractor, "connect_ssh_client")
+
+    dummy_channel = dummy_com()
+    mocker.patch.object(
+        dummy_extractor.ssh_client, "invoke_shell", return_value=dummy_channel
+    )
+    chan_send_mock = mocker.patch.object(dummy_channel, "send")
+    chan_recv_mock = mocker.patch.object(dummy_channel, "recv")
+    chan_close_mock = mocker.patch.object(dummy_channel, "close")
+
+    # Mock some other class functions
+    process_mock = mocker.patch.object(dummy_extractor, "process_ssh_output")
+    check_str_mock = mocker.patch.object(dummy_extractor, "check_ssh_str")
+
+    # Test working case
+    # First pass we get input, then recv nothing to break on second pass
+    recv_list = ["something", ""]
+    chan_recv_mock.side_effect = recv_list
+    check_str_mock.return_value = (0, [])
+    dummy_extractor.ssh_info_str = "something\n"
+
+    dummy_extractor.connect_thread_func()
+
+    ssh_client_mock.assert_called_once_with(
+        config_contents["test-fpga"]["ssh_user"], config_contents["test-fpga"]["ip"]
+    )
+    chan_send_mock.assert_has_calls(
+        [mocker.call("sudo su\n"), mocker.call(dummy_extractor.ssh_string)]
+    )
+    assert chan_recv_mock.call_count == len(recv_list)
+    chan_close_mock.assert_called_once()
+    chan_close_mock.reset_mock()
+    process_mock.assert_called_once()
+    check_str_mock.assert_called_once()
+
+    # Test recv fatal error
+    check_str_mock.return_value = (2, [])
+    recv_list = ["something", "another thing"]
+    chan_recv_mock.side_effect = recv_list
+    dummy_extractor.ssh_info_str = "something\n"
+
+    with pytest.raises(RuntimeError):
+        dummy_extractor.connect_thread_func()
+
+    chan_close_mock.assert_called_once()
+
+
+def test_connect(dummy_extractor, mocker):
+    """Test the IDExtractor's connect function
+
+       The start_mock always reports 0 calls, is it in a different thread?
+    """
+
+    # Don't actually create and start a thread
+    thread_mock = mocker.patch("threading.Thread")
+    # start_mock = mocker.patch("threading.Thread.start")
```

Comment:
Right, I left this in as more of a comment/discussion point.

I couldn't get the thread start to register, not sure if it's worth looking into or we just remove this code since we have good coverage anyways.

Reply:
Just saw your comment below, I'll remove the code.